{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGCTTCGGCATGTCCATTGAGCCCTTTAGT\n",
      "ACTAAGGGGCTCAATGGCCATGCTGAGGCT\n",
      "TGCTTCCGCATGGCCGTTGAGTCCCTTTGT\n",
      "GGCCTCGGCGTGACCATTTAGTCCTTTTGT\n",
      "CGCTTCGGCATGTCCGTTTAGGCCCTTCGT\n",
      "ACCAAAGGACTAAATGGACACGCAGAAGCA\n",
      "ACGAAAGGACTAAATGGCCACGCAGAGGCA\n",
      "TGCCTCAGCATGGCCATTGAGTCCCTTGGT\n",
      "AGCCTCCGCGTGGCCATTCAATCCCTTGGT\n",
      "ACTAAAGGTCTGAACGGTCACGCTGAGGCA\n",
      "ACCAAAGGTCTCAATGGTCATGCTGAAGCC\n",
      "ACCAAAGGCCTAAATGGGCATGCAGAGGCG\n",
      "GGCCTCGGCATGACCGTTCAGTCCTTTTGT\n",
      "ACTAAAGGATTGAATGGGCATGCAGAAGCG\n",
      "AGCCTCGGCATGTCCGTTTAAGCCTTTGGT\n",
      "AGCCTCGGCATGACCGTTCAAGCCTTTCGT\n",
      "ACCAAGGGGCTAAATGGGCACGCGGAGGCC\n",
      "TGCCTCCGCGTGGCCATTAAGCCCCTTGGT\n",
      "CGCCTCCGCGTGTCCGTTTAGTCCTTTGGT\n",
      "CGCCTCGGCGTGACCATTCAATCCCTTAGT\n"
     ]
    }
   ],
   "source": [
    "# C:\\Users\\wwsch\\anaconda3\\python.exe\n",
    "# Name: William Schlough (wschloug)\n",
    "# Group Members: Junmo, Mikey, Wenyu\n",
    "\n",
    "\n",
    "class peptide:\n",
    "    '''\n",
    "    class peptide contains all methods required to solve problem16 on rosalind.\n",
    "    The prompt is as follows:\n",
    "\n",
    "    Find substrings of a genome encoding a given amino acid sequence.\n",
    "    Given: A DNA string Text and an amino acid string Peptide.\n",
    "    Return: All substrings of Text encoding Peptide (if any such substrings exist).\n",
    "    Sample:\n",
    "    ATGGCCATGGCCCCCAGAACTGAGATCAATAGTACCCGTATTAACGGGTGA\n",
    "    MA\n",
    "\n",
    "    '''\n",
    "\n",
    "    '''\n",
    "    aa_to_dna is a re-organized dna codon table -- it is still a dictionary,\n",
    "    but it groups amino acid codons into lists.\n",
    "    '''\n",
    "    aa_to_dna = {'F': ['TTT', 'TTC'],\n",
    "                 'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "                 'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "                 'Y': ['TAT', 'TAC'],\n",
    "                 'C': ['TGT', 'TGC'],\n",
    "                 'W': ['TGG'],\n",
    "                 'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "                 'H': ['CAT', 'CAC'],\n",
    "                 'Q': ['CAA', 'CAG'],\n",
    "                 'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "                 'I': ['ATT', 'ATC', 'ATA'],\n",
    "                 'M': ['ATG'],\n",
    "                 'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "                 'N': ['AAT', 'AAC'],\n",
    "                 'K': ['AAA', 'AAG'],\n",
    "                 'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "                 'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "                 'D': ['GAT', 'GAC'],\n",
    "                 'E': ['GAA', 'GAG'],\n",
    "                 'G': ['GGT', 'GGC', 'GGA', 'GGG']\n",
    "                 }\n",
    "\n",
    "    def __init__(self, sequence, peptide):\n",
    "        '''\n",
    "        declare necessary vars\n",
    "        '''\n",
    "\n",
    "        self.seq = sequence\n",
    "        self.pep = peptide\n",
    "\n",
    "    def reverseSeq(self, sequence):\n",
    "        '''\n",
    "\n",
    "        used to find reverse compliment -- method taken from bme 160, very useful\n",
    "        returns the reverse compliment\n",
    "\n",
    "        https://www.w3schools.com/python/ref_string_translate.asp\n",
    "        https://www.programiz.com/python-programming/methods/string/maketrans\n",
    "\n",
    "        '''\n",
    "\n",
    "        #(sequence.translate(str.maketrans(\"ATCG\", \"TAGC\"))[::-1])\n",
    "        return sequence.translate(str.maketrans(\"ATCG\", \"TAGC\"))[::-1]\n",
    "\n",
    "    def genDNA(self, peptide):\n",
    "        '''\n",
    "        generates dna sequence from peptide sequence\n",
    "\n",
    "        https://stackoverflow.com/questions/231767/what-does-the-yield-keyword-do\n",
    "        '''\n",
    "\n",
    "        seq = []\n",
    "\n",
    "        # never starts true (out peptide will be at least 2 letters) but if peptide seq was hypothetically 1\n",
    "        if len(peptide) == 1:\n",
    "            # for codon in amino acid to dna dictionary matching peptide\n",
    "            for codon in self.aa_to_dna[peptide]:\n",
    "                # append it to seq list\n",
    "                seq.append(codon)\n",
    "\n",
    "        # otherwise (peptide is >1 letters)\n",
    "        else:\n",
    "            prefix = peptide[:-1] # cut all the way up to last element\n",
    "            suffix = peptide[-1] # last element\n",
    "\n",
    "            # for each codon that matches the suffix peptide\n",
    "            for codon in self.aa_to_dna[suffix]:\n",
    "                # for each dna generations of the prefix (remaining part of peptide)\n",
    "                for p in self.genDNA(prefix): # call genDNA on prefix (peptide sequence minus last peptide)\n",
    "                    # creates prefix + all suffix generations, but also recursively calls on prefix (since we\n",
    "                    # can [will] have a prefix of greater than 1 element, as seen in given file from rosalind)\n",
    "                    # for sample MA -- M only ATG, but A has 4 codon, we need to add all 4\n",
    "                    seq.append(p + codon)\n",
    "\n",
    "        # return seq list, which will contain every single formation of peptide sequence (not including reverseSeq)\n",
    "        return seq\n",
    "\n",
    "    def substring(self, sequence, peptide):\n",
    "        '''\n",
    "        generate substrings of dna sequence\n",
    "\n",
    "        \"We say that a DNA string Pattern encodes an amino acid string Peptide if the RNA string transcribed\n",
    "        from either Pattern or its reverse complement Pattern translates into Peptide.\"\n",
    "        '''\n",
    "\n",
    "        n = len(sequence)\n",
    "        k = len(peptide)*3\n",
    "\n",
    "        encoding = []\n",
    "        substrings = []\n",
    "\n",
    "        # for each sequence in generated DNA from peptide\n",
    "        for seq in self.genDNA(peptide):\n",
    "            # we also want the reverse sequence, add them both to a list\n",
    "            rs = self.reverseSeq(seq)\n",
    "            encoding.append(seq)\n",
    "            encoding.append(rs)\n",
    "            # sample MA encoding: ['ATGGCT', 'AGCCAT', 'ATGGCC', 'GGCCAT', 'ATGGCA', 'TGCCAT', 'ATGGCG', 'CGCCAT']\n",
    "\n",
    "        # range of (len(seq) - len(pep)*3 + 1) -- sample == (52 - 6 + 1)\n",
    "        # range(n) also works, dont notice time difference in program runtime\n",
    "        # ensures that we are calculating across proper range (peptide = codon = 3 nuc)\n",
    "        for i in range(n):\n",
    "            # cut substring from i up to i+6 (sample)\n",
    "            substring = sequence[i:i+k]\n",
    "\n",
    "            # if substring already seen in encoding\n",
    "            if substring in encoding:\n",
    "                # append it to substrings list\n",
    "                substrings.append(substring)\n",
    "\n",
    "        #return substrings list\n",
    "        return substrings\n",
    "\n",
    "    def print(self):\n",
    "        '''\n",
    "        quick method to print all substrings of a genome encoding a given amino acid sequence\n",
    "        '''\n",
    "\n",
    "        substrings = self.substring(self.seq, self.pep)\n",
    "\n",
    "        # for substring in substrings list\n",
    "        for substring in substrings:\n",
    "            # print substring\n",
    "            print(substring)\n",
    "\n",
    "\n",
    "def main(inFile = None):\n",
    "    '''\n",
    "    main() opens input file and reads it, then instantiates sequence into class peptide and calls print()\n",
    "    which calls the remaining methods to create all substrings of a genome encoding a given AA sequence\n",
    "\n",
    "    in main(inFile = '') you can input your own file for testing, as long as format follows 2 lines, with\n",
    "    first line being given AA sequence, and second line being the peptide\n",
    "\n",
    "    '''\n",
    "\n",
    "    count = 0\n",
    "    seq = ''\n",
    "    pep = ''\n",
    "\n",
    "    f = open(inFile)\n",
    "    for line in f:\n",
    "        if count == 0:\n",
    "            seq = line[0:].strip()\n",
    "        if count == 1:\n",
    "            pep = line[0:].strip()\n",
    "        count += 1\n",
    "\n",
    "    Pep = peptide(seq, pep)\n",
    "    Pep.print()\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main(inFile='rosalind_ba4b.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "__CODE MARKDOWN CELL__\n",
    "\n",
    "problem16 contains class peptide, along with a main\n",
    "\n",
    "class peptide contains:\n",
    "\n",
    "def __ __init__ __(self, sequence, peptide) -- initializes seq and pep\n",
    "\n",
    "def __reverseSeq__(self, sequence) -- used to find reverse compliment\n",
    "\n",
    "def __genDNA__(self, peptide) -- generates dna sequence from peptide sequence\n",
    "\n",
    "def __substring__(self, sequence, peptide) -- generate substrings of dna sequence\n",
    "\n",
    "def __print__(self) -- print all substrings of a genome encoding a given amino acid sequence\n",
    "\n",
    "__main__() reads from an input file and calls necessary methods\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "INSPECTION MARKDOWN CELL\n",
    "\n",
    "16. Junmo -- Nice work with your code! When generating dna sequence from peptide sequence, assigning a prefix as previous\n",
    "amino acids and a suffix as new amino acid to be added seems a brilliant idea.\n",
    "The code looks neat and the output looks good, too.\n",
    "If you add markdown cells when writing it into a notebook file later, it would be good.\n",
    "\n",
    "Inspector: Michael Collins\n",
    "Inspectee: William S\n",
    "\n",
    "Problem 16:\n",
    "Interesting that you chose to work in the DNA space instead of the amino acid space, i guess it is a pain to\n",
    "isolate codons from the sequence. No idea what's better but its nice to see new solutions. Other than that it\n",
    "looks great! Needs the markdown cell."
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}