{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "147-101-87-103-99-156-115-131-137-137 137-131-115-156-99-103-87-101-147-137 87-103-99-156-115-131-137-137-147-101 101-87-103-99-156-115-131-137-137-147 87-101-147-137-137-131-115-156-99-103 115-131-137-137-147-101-87-103-99-156 99-156-115-131-137-137-147-101-87-103 99-103-87-101-147-137-137-131-115-156 131-115-156-99-103-87-101-147-137-137 103-87-101-147-137-137-131-115-156-99 115-156-99-103-87-101-147-137-137-131 101-147-137-137-131-115-156-99-103-87 156-99-103-87-101-147-137-137-131-115 156-115-131-137-137-147-101-87-103-99 131-137-137-147-101-87-103-99-156-115 147-137-137-131-115-156-99-103-87-101 137-147-101-87-103-99-156-115-131-137 137-137-131-115-156-99-103-87-101-147 137-137-147-101-87-103-99-156-115-131 103-99-156-115-131-137-137-147-101-87\n"
     ]
    }
   ],
   "source": [
    "# C:\\Users\\wwsch\\anaconda3\\python.exe\n",
    "# Name: William Schlough (wschloug)\n",
    "# Group Members: Junmo, Mikey, Wenyu\n",
    "\n",
    "\n",
    "class idealMatch:\n",
    "    '''\n",
    "    class idealMatch contains all methods required to calculate the cyclic peptide whose theoretical spectrum matches\n",
    "    the experimental spectrum\n",
    "    The prompt is as follows:\n",
    "    Given an ideal experimental spectrum, find a cyclic peptide whose theoretical spectrum matches the experimental spectrum.\n",
    "    Given: A collection of (possibly repeated) integers Spectrum corresponding to an ideal experimental spectrum.\n",
    "    Return: Every amino acid string Peptide such that Cyclospectrum(Peptide) = Spectrum (if such a string exists).\n",
    "    Sample:\n",
    "    0 113 128 186 241 299 314 427\n",
    "    '''\n",
    "\n",
    "    # mass list for aa\n",
    "    masses = [\n",
    "        57, 71, 87, 97, 99,\n",
    "        101, 103, 113, 114, 115,\n",
    "        128, 129, 131, 137, 147,\n",
    "        156, 163, 186\n",
    "    ]\n",
    "\n",
    "    def __init__(self, spectrum):\n",
    "        '''\n",
    "        initialize necessary vars\n",
    "        '''\n",
    "\n",
    "        self.spectrum = spectrum\n",
    "\n",
    "    def expand(self, peptides):\n",
    "        '''\n",
    "        define EXPAND(Peptides) as a new collection containing all possible extensions of peptides in Peptides\n",
    "        by a single amino acid mass.\n",
    "        '''\n",
    "\n",
    "        # for peptide in copy of peptides (do this so we can remove)\n",
    "        for peptide in peptides.copy():\n",
    "            # remove the peptide from original\n",
    "            peptides.remove(peptide)\n",
    "            # for mass in masses dictionary\n",
    "            for m in self.masses:\n",
    "                # if peptide (true)\n",
    "                if peptide:\n",
    "                    # create new peptide which is a junction of original mass + new mass from mass list\n",
    "                    # (does this for each entry)\n",
    "                    new_pep = peptide + '-' + str(m)\n",
    "                # otherwise (not peptide)\n",
    "                else:\n",
    "                    # new peptide is just set to a mass from mass list\n",
    "                    new_pep = str(m)\n",
    "                # add to peptide list whatever the new peptide was\n",
    "                peptides.add(new_pep)\n",
    "\n",
    "        return peptides\n",
    "\n",
    "\n",
    "    def calcMass(self, peptide):\n",
    "        '''\n",
    "        quick method to calculate mass of peptide\n",
    "        '''\n",
    "        # map will create a map obj with items from peptide, split at '-' if conjunction\n",
    "        # then, we can sum the items in the map obj to find the mass of the conjunction peptide\n",
    "        mass = sum(map(int, peptide.split('-')))\n",
    "        return mass\n",
    "\n",
    "    def cycloSpec(self, peptide):\n",
    "        '''\n",
    "        method to calculate (sorted) cyclospectrum of input peptide\n",
    "        utilizes map(), sum()\n",
    "\n",
    "        sample example:\n",
    "        [128, 113, 186]\n",
    "        128\n",
    "        113\n",
    "        186\n",
    "        sum1-2: 241\n",
    "        sum2-3: 299\n",
    "        residual: [186, 128]\n",
    "        residual sum (3-1): 314\n",
    "        '''\n",
    "\n",
    "        cs = [0]\n",
    "        # create map obj of integers of peptides split at '-',\n",
    "        # then create a list from map obj\n",
    "        comp = list(map(int, peptide.split('-')))\n",
    "        n = len(comp)\n",
    "        # search in range of 1 up to but not including n (sample == 1 up to but not including 3)\n",
    "        for k in range(1, n):\n",
    "            # search in range of 0 up to but not including n (sample == 0 up to but not including 3)\n",
    "            for i in range(n):\n",
    "                # if range1 + range2 less than total length of composition\n",
    "                if i + k <= n:\n",
    "                    # cut the comp from location i to location i+k and append it to our list\n",
    "                    cs.append(sum(comp[i:i+k]))\n",
    "                # else (range1 + range2 > total length of composition)\n",
    "                else:\n",
    "                    # residual set to i+k-n\n",
    "                    r = i + k - n\n",
    "                    cs.append(sum(comp[i:] + comp[:r]))\n",
    "\n",
    "        cs.append(self.calcMass(peptide))\n",
    "        cs.sort()\n",
    "        return cs\n",
    "\n",
    "    def linearSpec(self, peptide):\n",
    "        '''\n",
    "        similar to cycloSpec, method to calculate (sorted) linear spectrum of peptide\n",
    "        utilizes map(), sum()\n",
    "\n",
    "        sample example:\n",
    "        comp cut: [186, 128]\n",
    "        comp sum : 314\n",
    "        '''\n",
    "\n",
    "        cs = [0]\n",
    "        # create map obj of integers of peptides split at '-',\n",
    "        # then create a list from map obj\n",
    "        comp = list(map(int, peptide.split('-')))\n",
    "        n = len(comp)\n",
    "\n",
    "        # search across range of len(comp)\n",
    "        for k in range(n):\n",
    "            # search across range of (len(comp) + 1)\n",
    "            for i in range(n + 1):\n",
    "                # add a sum of comp cut from i to i+k\n",
    "                cs.append(sum(comp[i:i+k]))\n",
    "\n",
    "        cs.append(self.calcMass(peptide))\n",
    "        cs.sort()\n",
    "        return cs\n",
    "\n",
    "    def notConsistent(self, spectrum, peptide):\n",
    "        '''\n",
    "        method nonConsistent will return false if the subpeptide is consistent,\n",
    "        and true is the subpeptide isn't consistent\n",
    "\n",
    "        \"The key to our new algorithm is that every linear subpeptide of a cyclic peptide Peptide is consistent with\n",
    "        CYCLOSPECTRUM(Peptide). Thus, to solve the Cyclopeptide Sequencing Problem for Spectrum, we can safely ban\n",
    "        all peptides that are inconsistent with Spectrum from the growing set Peptides, which powers the\n",
    "        bounding step that we described above.\" -- textbook pg 196\n",
    "\n",
    "        '''\n",
    "\n",
    "        # generate linearSpec of peptide\n",
    "        peptide_spectrum = self.linearSpec(peptide)\n",
    "        # for each value in linear spectrum\n",
    "        for val in peptide_spectrum:\n",
    "            # if the val isn't in the spectrum, return true (meaning it wasn't consistent)\n",
    "            if val not in spectrum:\n",
    "                return True\n",
    "        # otherwise return false (meaning is was consistent)\n",
    "        return False\n",
    "\n",
    "    def cyclopeptideSequencing(self, spectrum):\n",
    "        '''\n",
    "        pseudocode taken from textbook/rosalind prompt, methods created are based off pseudocode\n",
    "        https://www.programiz.com/python-programming/methods/built-in/set\n",
    "\n",
    "        from pseudocde, will know i want expand(), caluclate-mass(), cyclospectrum(), not consistent(), also linear()\n",
    "        because it looks like not-consistent() will use the linear spectrum\n",
    "\n",
    "        '''\n",
    "\n",
    "        peptides = {''}\n",
    "        matches = set() # use set so we can use add()\n",
    "\n",
    "        # while peptides not empty\n",
    "        while peptides:\n",
    "            # call expand on peptides\n",
    "            peptides = self.expand(peptides)\n",
    "            # for each peptide in peptides (use copy)\n",
    "            for peptide in peptides.copy():\n",
    "                # if mass == parent mass (max(spectrum))\n",
    "                if self.calcMass(peptide) == max(spectrum):\n",
    "                    # if cyclospectrum of peptide == spectrum\n",
    "                    if self.cycloSpec(peptide) == spectrum:\n",
    "                        # add it as a match\n",
    "                        matches.add(peptide)\n",
    "                    # since we used a copy, its ok for us to remove form original\n",
    "                    peptides.remove(peptide)\n",
    "                # else if peptide not consistent with spectrum\n",
    "                elif self.notConsistent(spectrum, peptide):\n",
    "                    # since we used a copy, its ok for us to remove from original\n",
    "                    peptides.remove(peptide)\n",
    "        # return all matches\n",
    "        return matches\n",
    "\n",
    "    def print(self):\n",
    "        '''\n",
    "        quick method to print all elements returned form cyclopeptideSequencing() after calling it with spectrum\n",
    "        '''\n",
    "\n",
    "        print(*self.cyclopeptideSequencing(self.spectrum))\n",
    "\n",
    "\n",
    "def main(inFile = None):\n",
    "    '''\n",
    "    learning how to use map() -- seems like a super useful function that i can use throughout\n",
    "    this program, along with in future assignments (hopefully :P)\n",
    "    https://www.geeksforgeeks.org/python-map-function/\n",
    "    https://realpython.com/python-map-function/\n",
    "    '''\n",
    "\n",
    "    spectrum = list()\n",
    "    f = open(inFile)\n",
    "    for line in f:\n",
    "        spectrum = list(map(int, line.split()))\n",
    "\n",
    "    # for sample, spectrum being taken into idealMatch looks like this:\n",
    "    # [0, 113, 128, 186, 241, 299, 314, 427]\n",
    "\n",
    "    pep = idealMatch(spectrum)\n",
    "    pep.print()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main(inFile='rosalind_ba4e.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "__CODE MARKDOWN CELL__\n",
    "\n",
    "problem18 contains class idealMatch, along with a main\n",
    "\n",
    "class idealMatch contains:\n",
    "\n",
    "def __ __init__ __(self, spectrum) -- initializes spectrum\n",
    "\n",
    "def __expand__(self, peptides) -- define EXPAND(Peptides) as a new collection containing all possible extensions of peptides in Peptides by a single amino acid mass.\n",
    "\n",
    "def __calcMass__(self, peptide) -- calculate mass of peptide\n",
    "\n",
    "def __cycloSpec__(self, peptide) -- calculate (sorted) cyclospectrum of input peptide\n",
    "\n",
    "def __linearSpec__(self, peptide) -- calculate (sorted) linear spectrum of peptide\n",
    "\n",
    "def __notConsistent__(self, spectrum, peptide) -- every linear subpeptide of a cyclic peptide Peptide is consistent with CYCLOSPECTRUM(Peptide). can safely ban\n",
    "all peptides that are inconsistent with Spectrum from the growing set Peptides\n",
    "\n",
    "def __cyclopeptideSequencing__(self, spectrum) -- pseudocode taken from textbook, implements branch + bound -- branching step is expand, bound step is notConsistent (including linearSpec)\n",
    "\n",
    "__main__() reads from an input file and calls necessary methods"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "INSPECTION MARKDOWN CELL\n",
    "\n",
    "18. Junmo -- Well written codes with comments and docstrings to easily explain about codes.\n",
    "Using .copy() was efficient to preserve the original when using remove function.\n",
    "I might consider using it next time instead of assigning a new variable.\n",
    "The output looks fine and well done with generating a pseudocode into a real code.\n",
    "As mentioned, markdown cells would be required.\n",
    "\n",
    "Inspector: Michael Collins\n",
    "Inspectee: William S\n",
    "\n",
    "Problem 18:\n",
    "I like the idea of having a notConsistent() method, looks very clean. I used list compressions but its kinda ugly.\n",
    "I thought it was easier to represent the peptides as lists of masses(ints) rather than a string with dashes, that way you don't\n",
    "have to keep translating it. I like the print function being worked into the class, imma start doing that.\n",
    "Overall looks great! Dont forget the markdown cell."
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}