{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 18\n",
    "Given an ideal experimental spectrum, find a cyclic peptide whose theoretical spectrum matches the experimental spectrum.\n",
    "\n",
    "Given: A collection of (possibly repeated) integers Spectrum corresponding to an ideal experimental spectrum.\n",
    "\n",
    "Return: Every amino acid string Peptide such that Cyclospectrum(Peptide) = Spectrum (if such a string exists).\n",
    "\n",
    "Functions:\n",
    "- Expand(Peptides)：a new collection containing all possible extentions of peptides in Peptides by a single amino acid.\n",
    "- ParentMass(Spectrum): the peptide’s mass in mass spectrometry experiments\n",
    "- Mass(Peptide): the largest mass in the spectrum.\n",
    "\n",
    "Algorithm:\n",
    "\n",
    "    CYCLOPEPTIDESEQUENCING(Spectrum):\n",
    "        Peptides ← a set containing only the empty peptide\n",
    "            while Peptides is nonempty\n",
    "                Peptides ← Expand(Peptides)\n",
    "                for each peptide Peptide in Peptides\n",
    "                    if Mass(Peptide) = ParentMass(Spectrum)\n",
    "                        if Cyclospectrum(Peptide) = Spectrum\n",
    "                            output Peptide\n",
    "                        remove Peptide from Peptides\n",
    "                    else if Peptide is not consistent with Spectrum\n",
    "                        remove Peptide from Peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "class generatePeptide():\n",
    "    \"\"\"\n",
    "    A class to produce cyclic peptides whose theoretical spectrum matches the experimental spectrum\n",
    "\n",
    "    ...\n",
    "\n",
    "    Attributes\n",
    "    ----------\n",
    "    experimentSpec : list\n",
    "        experimental spectrum\n",
    "    masses : list\n",
    "        the masses of standard amino acid\n",
    "\n",
    "    Methods\n",
    "    -------\n",
    "    expand(Peptides):\n",
    "        all extentions of peptides whose total mass in experimental spectrum by a single amino acid\n",
    "    Cyclospectrum(Peptides):\n",
    "        generate Spectrum for the cyclic peptide\n",
    "    cycloPeptideSequencing():\n",
    "        return the Peptides with identical spectrum to the experimental spectrum\n",
    "    \"\"\"\n",
    "    def __init__(self,infile):  \n",
    "        '''\n",
    "        contructor: saves attributes \n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "            infile : file name\n",
    "                \n",
    "        '''\n",
    "        with open(infile) as file:\n",
    "            #turn the element of experiment spectrum to int and sort the list\n",
    "            self.experimentSpec=sorted([int(i) for i in file.readline().strip().split()]) \n",
    "        #this is all the mass of standard amino acid\n",
    "        masses=[57, 71, 87, 97, 99, 101, 103, 113, 114, 115, 128, 129, 131, 137, 147, 156, 163, 186]\n",
    "        #filter the masses and remain those are consistent to experiment spectrum\n",
    "        self.masses=[i for i in masses if i in self.experimentSpec]\n",
    "    \n",
    "    def expand(self,Peptides):\n",
    "        \"\"\"\n",
    "        Generate the SubPeptide for a peptide\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        Peptides:list\n",
    "            The current peptides\n",
    "        \n",
    "        Return\n",
    "        ----------\n",
    "        Peptides:list\n",
    "            New peptides with one amino acid extended\n",
    "        \"\"\"\n",
    "        #----------------copy every peptide for 18 times so that each of them extend an amino acid--------------#\n",
    "        for i in list(Peptides):\n",
    "            j=1\n",
    "            while j<len(self.masses): #keep copying every element for 18 times\n",
    "                Peptides.append(list(i)) #the reason I use list(i) is that to avoid append an element with same id\n",
    "                j+=1\n",
    "        Peptides.sort() #sort the Peptides to make sure the same copies of element are put together\n",
    "        #----------------copy every peptide for 18 times so that each of them extend an amino acid--------------#\n",
    "        \n",
    "        #--------------------------here we extend 18 amino acids to each of the subpeptides---------------------#\n",
    "        removePep=[]\n",
    "        for start in range(0,len(Peptides),len(self.masses)):#every 18 elements(len(self.masses)) in peptides are in a group\n",
    "            for i in range(start,start+len(self.masses)): #start:start position for a new group\n",
    "        #------we append those inconsistent next amino acid to removePep and then remove them from Peptides-----#\n",
    "                temp=Peptides[i][:] #here we assign to next amino acid to extend to temp\n",
    "                temp.append(self.masses[i%len(self.masses)]) #temp is the peptide with the amino acid we're going to extend\n",
    "                if sum(temp) not in self.experimentSpec:#if the sum of masses in temp not in self.experimentSpec\n",
    "                    removePep.append(Peptides[i])\n",
    "                    continue\n",
    "                else: #temp is consistent\n",
    "                    Peptides[i].append(self.masses[i%len(self.masses)])\n",
    "        #------we append those inconsistent next amino acid to removePep and then remove them from Peptides-----#\n",
    "        \n",
    "        #--------------------------here we extend 18 amino acids to each of the subpeptides---------------------#\n",
    "        for i in removePep:#we are removing those inconsistent peptide from Peptides\n",
    "            Peptides.remove(i)\n",
    "        return Peptides\n",
    "    \n",
    "    def Cyclospectrum(self,Peptides):\n",
    "        \"\"\"\n",
    "        Generate the masses of subpeptides \n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        Peptides:list\n",
    "            The current peptides\n",
    "        \n",
    "        Return\n",
    "        ----------\n",
    "        sorted(subPeptides):list\n",
    "            the sorted masses of Peptides\n",
    "        \"\"\"\n",
    "        Peptides.remove(0) #when we simulate the cyclic peptide, we first remove 0\n",
    "        cicPeptide=list(Peptides) #avoid the Peptides and cicPeptide share the same id\n",
    "        cicPeptide.extend(Peptides[:len(Peptides)-2]) #to mimic a cyclic peptide, we entend some amino acids to the end\n",
    "        subPeptides=[] #save all the subPeptides\n",
    "        for i in range(1,len(Peptides)): #generate subpeptides for all length\n",
    "            for j in range(len(Peptides)): #just like generate kmer, j is the start position\n",
    "                subPeptides.append(sum(cicPeptide[j:j+i]))\n",
    "        subPeptides.extend([0,sum(Peptides)]) #we finally add the complete peptide to the list\n",
    "        return sorted(subPeptides) \n",
    "    \n",
    "    def cycloPeptideSequencing(self):\n",
    "        \"\"\"\n",
    "        Produce cyclic peptides whose theoretical spectrum matches the experimental spectrum\n",
    "\n",
    "        Return\n",
    "        ----------\n",
    "        candidates:list\n",
    "            the possible cyclic peptides\n",
    "        \"\"\"\n",
    "        Peptides=[[0]] #Peptides start with 0\n",
    "        candidates=[] #this is a list saving peptides whose spectrum identical to the experimental one\n",
    "        while Peptides: # if the Peptides not empty\n",
    "            Peptides=self.expand(Peptides) #call the expand function to expand amino acid\n",
    "            for i in list(Peptides): \n",
    "                if sum(i)==max(self.experimentSpec): #if Mass(Peptide) = ParentMass(Spectrum)\n",
    "                    if self.Cyclospectrum(i)==self.experimentSpec: #if the spectrums are the same\n",
    "                        candidates.append('-'.join(map(str,i)))# format the output\n",
    "                        Peptides.remove(i)\n",
    "        return candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(infile):\n",
    "    '''\n",
    "    Find cyclic peptides here\n",
    "    \n",
    "    Parameters\n",
    "        ----------\n",
    "        infile : str \n",
    "            the filename of the fasta file\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        STDOUT\n",
    "    '''\n",
    "    Peptide=generatePeptide(infile) #instantiation\n",
    "    Mass=Peptide.cycloPeptideSequencing()\n",
    "    with open('output.txt','w') as file:\n",
    "        for i in Mass:\n",
    "            print(i,end='\\t',file=file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run the program here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    main('rosalind_ba4e.txt')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
