{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# C:\\Users\\wwsch\\anaconda3\\python.exe\n",
    "# Name: William Schlough (wschloug)\n",
    "# Group Members: Junmo, Mikey, Wenyu\n",
    "\n",
    "\n",
    "########################################################################\n",
    "# FastAreader\n",
    "# provided by Dr. B\n",
    "########################################################################\n",
    "import sys\n",
    "\n",
    "class FastAreader :\n",
    "\n",
    "    def __init__ (self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "\n",
    "        self.fname = fname\n",
    "\n",
    "    def doOpen (self):\n",
    "        if self.fname is '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta (self):\n",
    "\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# CommandLine\n",
    "# provided by Dr. B\n",
    "########################################################################\n",
    "\n",
    "class CommandLine() :\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse,\n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination exception Usage.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    available within the .args attribute of an object instantiated from CommandLine.\n",
    "    For example, if thisCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then thisCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "\n",
    "    '''\n",
    "\n",
    "    def __init__(self, inOpts=None) :\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "\n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(description = 'Program prolog - a brief description of what this thing does',\n",
    "                                              epilog = 'Program epilog - some other stuff you feel compelled to say',\n",
    "                                              add_help = True, #default is True\n",
    "                                              prefix_chars = '-',\n",
    "                                              usage = 'python randomizedMotifSearch.py -i int -k int -p float < input.fa > output.out'\n",
    "                                              )\n",
    "\n",
    "        # options as specified by assignment\n",
    "        self.parser.add_argument('-i', type=int, action='store', help='number of iterations in integer format')\n",
    "        self.parser.add_argument('-k', type=int, action='store', help='motif length in integer format')\n",
    "        self.parser.add_argument('-p', type=float, action='store', help='pseudo-count in float format')\n",
    "\n",
    "        if inOpts is None :\n",
    "            self.args = self.parser.parse_args()\n",
    "        else :\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "\n",
    "class Usage(Exception):\n",
    "    '''\n",
    "    Signal a Usage error, evoking a usage statement and eventual exit when raised.\n",
    "    '''\n",
    "    def __init__(self, msg):\n",
    "        self.msg = msg"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final sequence consensus: AAAAACTTAAAAA\n",
      "Final entropy score: 8.805314402499295\n"
     ]
    }
   ],
   "source": [
    "########################################################################\n",
    "# The Bioinformatics Class --> randomizedMotifSearch\n",
    "# Here is where most of your solution will end up.\n",
    "#\n",
    "########################################################################\n",
    "\n",
    "import random  # needed for randint\n",
    "import math  # needed for calculating entropy function (log2)\n",
    "\n",
    "class consensusCRISPR:\n",
    "\n",
    "    '''\n",
    "\n",
    "    class consensusCRISPR contains methods that are used to calculate a consensus motif sequence and entropy score for\n",
    "    a given set of sequences, which in our case are housed in a fastA file.\n",
    "\n",
    "    build the thing that needs doing\n",
    "\n",
    "    things i want // overarching structure ----------------------------------------------------------\n",
    "\n",
    "    initialization of user inputs   -- dna sequence, kmer size, pseudocount size, run size\n",
    "    generate random motifs          -- can use random.() python library\n",
    "    generate counts from motifs     -- include pseudocounts\n",
    "    generate profile from counts    -- probability distribution -- DON'T NEED THIS\n",
    "    generate most probable motifs from profile distribution\n",
    "    generate entropy score          -- use equation discussed in class, textbook, and on assignment webpage\n",
    "    generate a consensus            -- can call main algorithm from this function\n",
    "    main algorithm found in textbook -- describing randomized motif search\n",
    "\n",
    "\n",
    "\n",
    "    '''\n",
    "    def __init__(self, DNA, kmer, pseudoCount, runs):\n",
    "        '''\n",
    "\n",
    "        DNA: list of seq taken from fastAreader --> LIST\n",
    "        kmer: kmer (motif) length --> INT\n",
    "        pseudoCount: psuedo-count --> FLOAT\n",
    "        runs: number of iterations that we will run RMF --> INT\n",
    "        bestMotifs: stores the most probable motifs\n",
    "        highScore: global int\n",
    "\n",
    "        '''\n",
    "\n",
    "        self.dna = DNA\n",
    "        self.k = kmer\n",
    "        self.psc = pseudoCount\n",
    "        self.i = runs\n",
    "\n",
    "        self.bestMotifs = []\n",
    "        self.highScore = 1000000000000000000\n",
    "\n",
    "    def randomMotifs(self):\n",
    "        '''\n",
    "\n",
    "        take in -- list of sequences\n",
    "        produce -- randomly generated motifs of length k (kmer), specified by user\n",
    "\n",
    "        https://www.w3schools.com/python/ref_random_randint.asp\n",
    "        \"The randint() method returns an integer number selected element from the specified range.\"\n",
    "\n",
    "        '''\n",
    "\n",
    "        randomMotifs = [] # randomMotifs is a list that will be filled with randomly generated motifs of length k\n",
    "\n",
    "        for sequence in self.dna: # for each sequence\n",
    "            startP = random.randint(0, len(sequence) - self.k) # create a random starting point (startP)\n",
    "            randomMotifs.append(sequence[startP: startP + self.k]) # append a cut from startP up to startP + length of kmer\n",
    "\n",
    "        return randomMotifs # return list randomMotifs --> we now have a bunch of random motifs (of length k) to use\n",
    "\n",
    "    def counts(self, motifs):\n",
    "        '''\n",
    "\n",
    "        take in -- list of motifs\n",
    "        produce -- dictionary containing counts of nucs in positions across motif\n",
    "\n",
    "        '''\n",
    "\n",
    "        # dictionary of lists -- simulating matrix\n",
    "        countDictionary = {'A': [], 'T': [], 'G': [], 'C': []} # countDictionary will host all counts of ATGC\n",
    "\n",
    "        # we utilize our pseudocount variable here to increment the counts\n",
    "        inc = self.psc\n",
    "\n",
    "        for i in range(self.k): # for i in range 0 up to length of kmer\n",
    "            columnX = [column[i] for column in motifs]\n",
    "\n",
    "            countDictionary['A'].append(columnX.count('A') + inc) # add pseudocount\n",
    "            countDictionary['T'].append(columnX.count('T') + inc) # add pseudocount\n",
    "            countDictionary['G'].append(columnX.count('G') + inc) # add pseudocount\n",
    "            countDictionary['C'].append(columnX.count('C') + inc) # add pseudocount\n",
    "\n",
    "        #print(\"count dictionary\")\n",
    "        #print(countDictionary)\n",
    "        return countDictionary\n",
    "\n",
    "    def motif(self, counts):\n",
    "        '''\n",
    "\n",
    "        take in -- profile matrix for set of motifs\n",
    "        produce -- matrix of most probable motifs\n",
    "\n",
    "        https://www.w3schools.com/python/ref_keyword_none.asp\n",
    "        \"The None keyword is used to define a null value, or no value at all.\n",
    "        None is not the same as 0, False, or an empty string.\n",
    "        None is a data type of its own (NoneType) and only None can be None.\"\n",
    "\n",
    "        https://stackoverflow.com/questions/521674/initializing-a-list-to-a-known-number-of-elements-in-python\n",
    "\n",
    "        '''\n",
    "\n",
    "        # new motif matrix\n",
    "        # getting error \"list assignment index out of range\" unless I use [None] instead of []\n",
    "        # i try [1] and it also seems to work but for now I stick with none\n",
    "        motifMatrix = [None]*len(self.dna) # none inside []? trying to create a list with an element per element in dnaseq\n",
    "\n",
    "        for i in range(len(self.dna)):  # search across length of dna\n",
    "            sequence = self.dna[i]\n",
    "            highProb = 0  # initialize variable for use of highest probability for each motif\n",
    "\n",
    "            for x in range(len(sequence) - self.k + 1):  # search across range of sequence minus kmer size (+1 necessary to retain proper range)\n",
    "                currMotif = sequence[x: x + self.k]  # store current motif that we are viewing\n",
    "                probability = 1  # set probability\n",
    "\n",
    "                for y in range(self.k): # search across length of kmer\n",
    "                    nuc = currMotif[y]  # for each nucleotide base in the current motif of length kmer\n",
    "                    probability *= counts[nuc][y] # generate a probability of the nuc at its location using profile matrix\n",
    "\n",
    "                if probability > highProb: # starts true (we set highProb to 0)\n",
    "                    motifMatrix[i] = currMotif # keep this motif since its probability was higher than our highest probability!\n",
    "                    highProb = probability # set our highest probability to our new highest probability\n",
    "\n",
    "        #print(\"motif matrix\")\n",
    "        #print(motifMatrix)\n",
    "        return motifMatrix\n",
    "\n",
    "    def entropy(self, motif):\n",
    "        '''\n",
    "\n",
    "        takes in -- motif matrix, which then has counts() called on it\n",
    "        produce -- entropy score for motifs\n",
    "\n",
    "        '''\n",
    "\n",
    "        profile = self.counts(motif) # call profile on motif-input\n",
    "        score = 0 # base score temp var\n",
    "\n",
    "        totalCount = sum([seq[0] for seq in profile.values()]) # we need a total count of all values in the sequence (i.e total count of ALL ATGC)\n",
    "\n",
    "        for nuc in profile.keys(): # search across each nucleotide\n",
    "            for pos in profile[nuc]: # search across each position in matrix\n",
    "                if pos != 0: # we only look for positions with values > 0\n",
    "                    score += -1*((pos/totalCount) * math.log2(pos/totalCount)) # accumulate score using summation equation\n",
    "\n",
    "        #score = score * -1 # finish summation equation\n",
    "\n",
    "        #print(\"score\")\n",
    "        #print(score)\n",
    "        return score\n",
    "\n",
    "    def consensus(self):\n",
    "        '''\n",
    "\n",
    "        produce -- consensus motif (from best scoring motifs from randomizedMotifSearch())\n",
    "\n",
    "        https://www.w3schools.com/python/ref_func_max.asp\n",
    "\n",
    "        '''\n",
    "\n",
    "        self.randomizedMotifSearch() # call rms() to calculate best motifs\n",
    "        consensus = ''\n",
    "\n",
    "        #print(self.bestMotifs)\n",
    "        for i in range(self.k): # search across kmer\n",
    "            # creates a list with each position being the ith element from each motif (does this for each .bestMotif)\n",
    "            # example: we have 10 5-mers, each starting with A\n",
    "            # 1st list created would house 10 A's\n",
    "            # 2nd list created would house the second element from each motif, etc. (we would have 5 lists for this ex.)\n",
    "            nuc = [base[i] for base in self.bestMotifs]\n",
    "            #print(nuc)\n",
    "            # we can then find the base that occurs the most using max() on the list, which will determine the base\n",
    "            # that should be added to our consensus sequence\n",
    "            # example continued: if there were 10 A's in the 1st list, then we add A to our consensus sequence\n",
    "            # if there were 6 A's and 4 T's in the 2nd list, then we add another A to our consensus sequence, etc.\n",
    "            consensus += max(nuc, key=nuc.count)\n",
    "\n",
    "\n",
    "        #print(\"consensus\")\n",
    "        #print(consensus)\n",
    "        return consensus # return consensus sequence\n",
    "\n",
    "\n",
    "    def randomizedMotifSearch(self):\n",
    "        '''\n",
    "\n",
    "        produce -- best scoring motifs\n",
    "\n",
    "        pseudocode from textbook is as follows ----------------\n",
    "\n",
    "        randomly select k-mers motifs\n",
    "        bestMotifs <-- motifs\n",
    "        while true:\n",
    "            profile\n",
    "            motifs\n",
    "\n",
    "            if SCORE(motifs) < SCORE(bestMotifs)\n",
    "                bestMotifs <-- motifs\n",
    "            else\n",
    "                return bestMotifs\n",
    "\n",
    "\n",
    "        tips from dr b. -----\n",
    "        \"Check the way you are comparing final profile scores across trajectories.\n",
    "        Remember that we want the best score and the associated consensus that came from that best score.\n",
    "        This will not likely be from the last iteration (trajectory) that we ran\"\n",
    "\n",
    "        '''\n",
    "\n",
    "        for i in range(self.i): # search over all iterations, -i=INT defined by user in commandLine\n",
    "            bestMotifs = self.randomMotifs() # set bestMotifs var to the random motifs we calculated with randomMotifs()\n",
    "\n",
    "            while True:\n",
    "                count = self.counts(bestMotifs) # create profile from random motifs\n",
    "                currMotif = self.motif(count) # calculate new motifs based on profile\n",
    "\n",
    "                currMotifScore = self.entropy(currMotif) # calculate entropy for current motifs\n",
    "                bestMotifScore = self.entropy(bestMotifs) # calculate entropy for previous motifs\n",
    "\n",
    "                if currMotifScore < bestMotifScore: # compare scoring, if current motifscore is less than best motif score\n",
    "                    bestMotifs = currMotif # set previous equal to current\n",
    "\n",
    "                    if currMotifScore < self.highScore: # if current score is less than global variable highScore\n",
    "                        self.highScore = currMotifScore # set global variable highScore equal to current score\n",
    "                        self.bestMotifs = bestMotifs # set best motif equal to bestMotif (which has been set to current motif)\n",
    "\n",
    "                else: # otherwise (we finished!)\n",
    "                    break # exit the loop\n",
    "\n",
    "        return self.bestMotifs # return our best motifs\n",
    "\n",
    "\n",
    "########################################################################\n",
    "# Main\n",
    "# Here is the main program\n",
    "# This is a place to instantiate objects, read and write some data for\n",
    "# those objects and maybe implement some of the command line options.\n",
    "########################################################################\n",
    "\n",
    "def main(myCommandLine=None):\n",
    "    '''\n",
    "    Implement the Usage exception handler that can be raised from anywhere in process.\n",
    "\n",
    "    '''\n",
    "\n",
    "\n",
    "    if myCommandLine is None:\n",
    "        myCommandLine = CommandLine()  # read options from the command line\n",
    "    else :\n",
    "        myCommandLine = CommandLine(myCommandLine) # interpret the list passed from the caller of main as the commandline.\n",
    "\n",
    "\n",
    "    seqL = []\n",
    "    myReader = FastAreader('combinedCrisprs.fa')\n",
    "    for head, seq in myReader.readFasta():\n",
    "        seqL.append(seq)\n",
    "\n",
    "    finalConsensus = consensusCRISPR(DNA=seqL,\n",
    "                                     kmer=myCommandLine.args.k,\n",
    "                                     pseudoCount=myCommandLine.args.p,\n",
    "                                     runs=myCommandLine.args.i)\n",
    "\n",
    "    # i built it so that consensus() runs in tandem with randomizedMotifSearch() --\n",
    "    # calling .consensus() will call randomizedMotifSearch(), which uses all other methods to return\n",
    "    # the best motifs, which consensus then sorts through to find the max occurrence of each nucleotide base,\n",
    "    # and then builds the final sequence. Additionally, highscore has been calculated, but we want it in a string\n",
    "    # format, which is why the second print line has the str() function.\n",
    "    print(\"Final sequence consensus: \" + finalConsensus.consensus())\n",
    "    print(\"Final entropy score: \" + str(finalConsensus.highScore))\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main(['-i=1000',\n",
    "          '-p=1',\n",
    "          '-k=13'])"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Code Markdown Cell\n",
    "__Schlough_William_randomizedMotifSearch.ipynb__ contains ------------- 3 classes - FastAreader, CommandLine, and consensusCRISPR - along with a main\n",
    "\n",
    "__FastAreader__ and __CommandLine__ were provided by Dr. B, and only CommandLine was slightly changed to fit the program requirements\n",
    "\n",
    "__consensusCRISPR__ contains methods that are used to calculate a consensus motif sequence and entropy score for a given set of sequences,\n",
    "which in our case are housed in a fastA file.\n",
    "\n",
    "def __ __init__ __(self, DNA, kmer, pseudoCount, runs) -- initialization of user inputs occurs here\n",
    "\n",
    "def __counts__(self, motifs) -- takes in a list of motifs and generates a count matrix (makes use of a dictionary of lists)\n",
    "\n",
    "def __motif__(self, counts) -- takes in a count matrix of motifs and returns the most probably motifs\n",
    "\n",
    "def __entropy__(self, motif) -- takes in a motif matrix (which has counts called on it) and determines the entropy score using equation described in textbook + in lecture\n",
    "\n",
    "def __consensus__(self) -- produces a consensus motif from the best scoring motifs which come from calling randomizedMotifSearch() (these two methods work in tandem)\n",
    "\n",
    "def __randomizedMotifSearch__(self) -- produce the best scoring motifs, based on the pseudocode found in the textbook\n",
    "\n",
    "__main__ -- for this assignment, we print the final sequence consensus, along with the final entropy score. to do this, I had to code the variables into main, but i believe that it is supposed to work with user input (even in jupyter notebook) so I will figure out how to do this for the next assignment"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "########################################################################\n",
    "# inspection markdown\n",
    "# done by: Junmo\n",
    "# Neat and well visualized code. Initial motif is random, I can get a profiling score and the consensus, too.\n",
    "# I like the way you iterate by making one more function.\n",
    "# But I got an entropy score of 16 from your code in which I think it's a little bit higher.\n",
    "#\n",
    "# WS -- when only using one of the \"\".fa files, the entropy score IS higher, but when running it with all of the\n",
    "# \"\".fa files combined, my entropy score is around 8.8, which is good when compared to others.\n",
    "#\n",
    "########################################################################\n",
    "# Inspector: Michael Collins\n",
    "# Instead of 'take in' and 'produce' in docstrings use 'input' and 'output', and include the respective datatypes\n",
    "#\n",
    "# In Motif mehtod, you can used probablility *= instead of probablility = probablilty *\n",
    "#\n",
    "# could use count based profile instead of probablity based profile since you don't calcualte realative entorpy very often\n",
    "#\n",
    "# WS -- changes were made based on recommendations, code was shortened over 10% by removing profile() method,\n",
    "# with same result still being achieved\n",
    "########################################################################\n",
    "# inspection markdown\n",
    "# done by: Wenyu Liang\n",
    "# I guess it is not neccesary to check whether current motifscore is less than best motif score (in line #347) because when\n",
    "# we use randomized motif search, the RE will keep becoming greater\n",
    "# line 304 using key function in max is cool!\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}